torch/nn/parallel/distributed.py

```diff
@@ -612,6 +625,172 @@ def train(self, mode=True):
+        # ensures that we keep the same order in join mode, such as when bucket
+        # order is rebuilt dynamically.
+        all_bucket_tensors = self.reducer.get_bucket_tensors()
+        for bucket_tensors in all_bucket_tensors:
+            # Joined processes contribute zero gradient. To keep the gradient
+            # consistent, the dividing factor is reduced by 1 for each joined
+            # process.
+            zero_tensors = [
+                torch.zeros_like(t) for t in bucket_tensors
+            ]
+            work = self.process_group.allreduce(zero_tensors)
+            allreduce_work.append(work)
+        for work in allreduce_work:
+            work.wait()
+
+    # Allreduces the used parameter mapping across ranks.
+    def _match_unused_params_allreduce(self):
+        locally_used_param_maps = self.reducer._get_local_used_maps()
+        self.process_group.allreduce(locally_used_param_maps)
+
+    @contextmanager
+    def join(self, enable=True):
+        r"""
+        A context manager to be used in conjunction with an instance of
+        :class:`torch.nn.parallel.distributed.DistributedDataParallel` to be
+        able to train with uneven inputs across participating processes.
+
+        This context manager will keep track of already-joined DDP processes,
+        and "shadow" the forward and backward passes by inserting collective
+        communication operations to match with the ones created by non-joined
+        DDP processes. This will ensure each collective call has a corresponding
+        call by already-joined DDP processes, preventing hangs or errors that
+        would otherwise happen when training with uneven inputs across
+        processes.
+
+        Once all DDP processes have joined, the context manager will broadcast
+        the model corresponding to the last joined process to all processes to
+        ensure the model is the same across all processes
+        (which is guaranteed by DDP).
+
+        To use this to enable training with uneven inputs across processes,
+        simply wrap this context manager around your training loop. No further
+        modifications to the model or data loading is required.
+
+        .. warning::
+            This module works only with the multi-process, single-device usage
+            of :class:`torch.nn.parallel.distributed.DistributedDataParallel`,
+            which means that a single process works on a single GPU.
+
+        ..warning::
+            This module currently does not support custom distributed collective
+            operations in the forward pass, such as ``SyncBatchNorm`` or other
+            custom defined collectives in the model's forward pass.
+
+        Args:
+            enable (bool): Whether to enable uneven input detection or not. Pass
+             in``enable=False`` to disable in cases where you know that inputs
+             are even across participating processes. Default is ``True``.
```

Comment:
hmm, why do we need this arg? If users care certain that inputs are even, will it work if they just don't use this ctx manager?

Reply:
This is similar to the autograd profiler: https://github.com/pytorch/pytorch/blob/master/torch/autograd/profiler.py#L297 where the idea is that users can easily turn the context manager on/off with the boolean flag, instead of having to make more code changes when iterating on their code.